File: example/1d-linear-convection/weno3/julia/03g/src/reconstructor/factory.jl
# src/reconstructor/factory.jl

module ReconstructorFactory

const _REGISTRY = Dict{String, Function}()

function register(name::String, ctor::Function)
    if haskey(_REGISTRY, name)
        @warn "重构器 '$name' 已注册，将被覆盖"
    end
    _REGISTRY[name] = ctor
end

function create(cfd::Any)
    config = cfd.config
    domain = cfd.domain
    scheme = lowercase(string(getfield_safe(config, :recon_scheme, "")))
    
    if scheme == ""
        error("必须先通过 with_reconstruction 设置重建格式！")
    end

    # 处理 WENO 默认命名
    if scheme == "weno"
        order = getfield_safe(config, :spatial_order, 5)
        scheme = "weno$(order)"
    end

    if !haskey(_REGISTRY, scheme)
        available = sort(collect(keys(_REGISTRY)))
        error("未注册的重构器: '$scheme'。可用选项: $available")
    end

    return _REGISTRY[scheme](cfd)  # ✅ 只传 cfd
end


# ✅ 手动注册（与你原有风格一致）
register("eno", cfd -> Main.EnoReconstructor(cfd))
register("weno3", cfd -> Main.Weno3Reconstructor(cfd))
register("weno5", cfd -> Main.Weno5Reconstructor(cfd))

# ---------------------- 辅助函数 ----------------------
function getfield_safe(obj, name::Symbol, default)
    return isdefined(obj, name) ? getfield(obj, name) : default
end

end  # module ReconstructorFactory

export ReconstructorFactory

File: example/1d-linear-convection/weno3/python-v1/02/src/reconstructor/factory.py
# reconstructor/factory.py
from core.registry import BaseFactory

class ReconstructorFactory:
    @staticmethod
    def create(config, domain):
        scheme = config.recon_scheme.lower()
        order = getattr(config, 'spatial_order', None)
        
        if scheme == "weno":
            if order is None:
                raise ValueError("使用 'weno' 时必须设置 config.spatial_order")
            scheme = f"weno{order}"
            
        print(f"ReconstructorFactory scheme={scheme}")
        
        print(f"scheme={scheme}")
        if scheme == "eno":
            if order is None:
                order = 3
            return BaseFactory.create_component('reconstructor', scheme, order, domain.ntcells)
        elif scheme == "weno3" or scheme == "weno5":
            # WENO3,WENO5无参数
            print(f"scheme == 'eno3' or scheme == 'weno5' scheme={scheme}")
            return BaseFactory.create_component('reconstructor', scheme)
        else:
            # 其他情况
            return BaseFactory.create_component('reconstructor', scheme, config)